src/main/java/edu/ucla/library/iiif/auth/services/DatabaseService.java

```diff
@@ -0,0 +1,102 @@
+import io.vertx.core.Vertx;
+import io.vertx.core.json.JsonObject;
+import io.vertx.pgclient.PgPool;
+import io.vertx.redis.client.Redis;
+import io.vertx.serviceproxy.ServiceProxyBuilder;
+
+/**
+ * A service for accessing the Hauth database.
+ */
+@ProxyGen
+@VertxGen
+public interface DatabaseService {
+
+    /**
+     * The event bus address that the service will be registered on, for access via service proxies.
+     */
+    String ADDRESS = DatabaseService.class.getName();
+
+    /**
+     * Creates an instance of the service.
+     *
+     * @param aVertx A Vert.x instance
+     * @param aConfig A configuration
+     * @return The service instance
+     */
+    static DatabaseService create(final Vertx aVertx, final JsonObject aConfig) {
+        return new DatabaseServiceImpl(aVertx, aConfig);
+    }
+
+    /**
+     * Creates an instance of the service proxy. Note that the service itself must have already been instantiated with
+     * {@link #create} in order for this method to succeed.
+     *
+     * @param aVertx A Vert.x instance
+     * @return A service proxy instance
+     */
+    static DatabaseService createProxy(final Vertx aVertx) {
+        return new ServiceProxyBuilder(aVertx).setAddress(ADDRESS).build(DatabaseService.class);
+    }
+
+    /**
+     * Closes the underlying connections required by this service.
+     *
+     * @return A Future that resolves once the connections have been closed
+     */
+    @ProxyClose
+    Future<Void> close();
+
+    /**
+     * Gets the "access level" of the item with the given identifier.
+     *
+     * @param aId The item identifier
+     * @return A Future that resolves to the access level once it's been fetched
+     */
+    Future<Integer> getAccessLevel(String aId);
+
+    /**
+     * Sets the given "access level" of the item with the given identifier.
+     *
+     * @param aId The item identifier
+     * @param aAccessLevel The access level to set for the item
+     * @return A Future that resolves once the access level has been set
+     */
+    Future<Void> setAccessLevel(String aId, int aAccessLevel);
+
+    /**
+     * Gets the "degraded allowed" for content hosted at the given origin.
+     *
+     * @param aOrigin The origin
+     * @return A Future that resolves to the degraded allowed once it's been fetched
+     */
+    Future<Boolean> getDegradedAllowed(String aOrigin);
+
+    /**
+     * Sets the given "degraded allowed" for content hosted at the given origin.
+     *
+     * @param aOrigin The origin
+     * @param aDegradedAllowed The degraded allowed to set for the origin
+     * @return A Future that resolves once the degraded allowed has been set
+     */
+    Future<Void> setDegradedAllowed(String aOrigin, boolean aDegradedAllowed);
+
+    /**
+     * Gets the underlying database connection pool.
+     *
+     * @return A Future that resolves to the underlying database connection pool
+     */
+    Future<PgPool> getConnectionPool();
```

Comment:
This is where I was unsure about breaking the encapsulation. You're providing an interface to connect to the database, `DatabaseService`, abstracting away the underlying db. It seems to me that, if you want to provide access to a db pool, it should be a wrapped pool, also abstracting away the underlying db rather than returning something specific to the db. The same is true for the cache client (though maybe we remove the redis stuff from the interface for now? I didn't want to pull out all the redis plumbing, but I don't know that we need it in the public interface at this point either).

I realize this is more work, also abstracting `PgConnectOptions` and other things in the pool's interface that are specific to pg. I don't think you'd need to abstract SqlClient and the generic sql things since they're not specific to the pg db.

This is an API design comment. There is nothing functionally wrong with breaking the encapsulation here. We can use a pg specific pool and it will take more development time to not break the model's encapsulation, which you could argue is not worth it since this is just an API for a single application and not something that we're releasing on its own or something that other applications would use. Just want to raise the question though...

Reply:
I don't love having getConnectionPool and getRedisClient in the interface. The reason I added them is because the tests testDbUserCount and testDbCacheConnection require that lower-level access. Are those tests still needed / could we remove them?

Reply:
I think we could remove them now that we're also testing the database through the newly added tests now.

Reply:
I did like the 8-bit art that came from the Redis LOLWUT, but I guess I can live without it 🙃